README.md
mllp
====

HL7's MLLP (Minimum Lower Layer Protocol) server implementation in Node.js.


Listen on predefined port for HL7 messages in format:

```
<VT>[HL7 Message]<FS><CR>
```

* [HL7 MLP minimul layer protocol defined](http://www.hl7standards.com/blog/2007/05/02/hl7-mlp-minimum-layer-protocol-defined/)
* [ACK message](http://www.hl7standards.com/blog/2007/02/01/ack-message-original-mode-acknowledgement/)


## Quick up and running guide

### Prerequisites

- Node.js (v0.10+) and NPM

```
# you need Node.js installed

# install dependencies and test
yarn
yarn test
```

## Usage

See `example.js`:

```ts
import { MllpServer } from "./mllp";

const server: MllpServer = new MllpServer('127.0.0.1', 1234);

server.listen()

// Subscribe to inbound messages
server.on('hl7', (data) => {
    console.log('received payload:', data);
});

// Send outbound messages
server.send('127.0.0.1', 4321, 'outbound-hl7-message', (err: any, ackData: any) => {
    // async callback code here
    console.log(`Sent message!\nerr: ${err}\nackData: ${ackData}`)
});
```

## Contributing

Contributors are welcome. See issues on [GitHub issues](https://github.com/amida-tech/mllp/issues)

## Release Notes

See release notes [here] (./RELEASENOTES.md)

## License

Licensed under [Apache 2.0](./LICENSE)


RELEASENOTES.md
# MLLP Release Notes

# v1.0.0 - December 12, 2014

This is the initial release of this library.

- Basic listener server compliant with HL7's MLLP standard
- Sends AA acknowledgement
- Access to HL7 messages/data via EventEmitter

# v2.0.0 - September 25 , 2018

- Set minimum node version to be 8
- Large message support

# v3.0.0 - November 4, 2021

- Updated to Typescript
- Updated minimum node version 

test/test.js
var assert = require('chai').assert;
var fs = require('fs');
var mllp = require('../index.js');
var net = require('net');

var VT = String.fromCharCode(0x0b);
var FS = String.fromCharCode(0x1c);
var CR = String.fromCharCode(0x0d);

describe('test server with client data exchange', function () {
    var hl7 = '';
    var server;

    before(function () {
        hl7 = fs.readFileSync('./test/fixtures/test.txt').toString().split('\n').join('\r');

        server = new mllp.MLLPServer('127.0.0.1', 1234);
    });

    describe('sending and receiving HL7 messages', function () {
        var error;
        var data;

        // Sending
        beforeEach(function (done) {
            server.send('127.0.0.1', 1234, hl7, function (err, ackData) {
                error = err;
                data = ackData;
                done();
            });
        });

        // Receiving
        it('receives an HL7 ACK message without error', function () {
            assert.equal(error, null);
            assert.equal(data, 'MSA|AA|Q335939501T337311002');
        });

        it('receives an HL7 message', function () {
            server.on('hl7', function (data) {
                assert.equal(hl7, data);
            });
        });
    });

    describe('sending HL7 message that errors', function () {
        var error;
        var data;

        // Sending
        beforeEach(function (done) {
            // port 9999 is bogus
            server.send('127.0.0.1', 9999, hl7, function (err, ackData) {
                error = err;
                data = ackData;
                done();
            });
        });

        // Receiving
        it('receives an error response', function () {
            assert.isNotNull(error);
            assert.equal(data, null);
        });
    });

    describe('not passing in specific host and port', function () {
        beforeEach(function () {
            server = new mllp.MLLPServer();
        });

        it('uses sane defaults', function (done) {
            new net.connect({
                // defaults that the library uses
                host: '127.0.0.1',
                port: 6969
            }, function () {
                // no need for assertion... this is the success callback
                // if it doesn't get invoked, the test will not pass.
                done();
            });
        });
    });
});

describe("sends a large message for data exchange", function () {
    var hl7Message = '';

    before(function () {
        hl7Message = fs.readFileSync('./test/fixtures/LargeA08.txt').toString().split('\n').join('\r');

        server = new mllp.MLLPServer('127.0.0.1', 1235);
    });

    describe("sending a large A08 Message and Receiving an Ack Back", function () {
        var ack, error;

        beforeEach(function (done) {
            server.send("127.0.0.1", 1235, hl7Message, function (err, ackData) {
                error = err;
                ack = ackData;
                done();
            });
        });

        it("receives a HL7 Message", function () {
            server.on('hl7', function (data) {
                assert.equal(hl7Message, data);
            });
        });
    });
});


src/test/test.ts
import { MllpServer } from '../mllp'
import { assert } from 'chai'
import * as net from 'net'

var fs = require('fs')

describe('MllpServer initialized with port and host', function () {
    let hl7: string = '';
    let server: MllpServer;

    before((done) => {
        hl7 = fs.readFileSync('./src/test/fixtures/test.txt').toString().split('\n').join('\r')

        server = new MllpServer('127.0.0.1', 1234)
        server.listen()
        done()
    });

    after((done)=>{
        server.close(done)
    })


    describe('should send and recieve HL7 messages', function () {
        let error: any;
        let data: any;

        // Sending
        beforeEach((done) => {
            server.send('127.0.0.1', 1234, hl7, (err: any, ackData: any) => {
                error = err;
                data = ackData;
                done();
            });
        });

        // Receiving
        it('receives an HL7 ACK message without error', () => {
            assert.equal(error, null);
            assert.equal(data, 'MSA|AA|Q335939501T337311002');
        });

        it('receives an HL7 message', () => {
            server.on('hl7', function (data) {
                assert.equal(hl7, data);
            });
        });
    });

    describe('should handle HL7 message that errors', function () {
        var error: any;
        var data: any;

        // Sending
        beforeEach((done) => {
            // port 9999 is bogus
            server.send('127.0.0.1', 9999, hl7, (err: any, ackData: any) => {
                error = err;
                data = ackData;
                done();
            });
        });

        // Receiving
        it('receives an error response', () => {
            assert.isNotNull(error);
            assert.equal(data, null);
        });
    });

    describe("should send a large A08 Message and recieve an ACK back", () => {
        let largeHl7 = fs.readFileSync('./src/test/fixtures/LargeA08.txt').toString().split('\n').join('\r')
        let server2: MllpServer

        var ack, error;

        beforeEach((done) => {
            server2 = new MllpServer('127.0.0.1', 1235)
            server2.listen()

            server.send("127.0.0.1", 1235, largeHl7, (err: any, ackData: any) => {
                error = err;
                ack = ackData;
                done();
            });
        });

        it("receives a HL7 Message", () => {
            server2.on('hl7', function (data) {
                assert.equal(largeHl7, data);
            });            
        });

        after((done)=>{
            server2.close(done)
        })
    
    });
});

describe('MllpServer without specific host and port', () => {
    let newServer: MllpServer

    it('should use sane defaults', (done) => {
        newServer = new MllpServer()
        newServer.listen()
        
        let client =  net.createConnection({
            // defaults that the library uses
            host: '127.0.0.1',
            port: 6969
        }, function () {
            client.end()
        });

        client.end(done)
    });

    after((done)=>{
        newServer.close(done)
    })
});

src/mllp.ts
import * as net from 'net'
import { TypedEmitter } from 'tiny-typed-emitter';

const hl7: any = require('hl7') 

/**
 * @constructor MllpServer
 * @param {string} host a resolvable hostname or IP Address
 * @param {integer} port a valid free port for the server to listen on.
 * @param {object} logger
 * 
 * @fires MLLPServer#hl7  
 * 
 * @example
 * var server = new MllpServer('hl7server.mydomain', 3333, console.log);
 * 
 * server.on('hl7', function(message) {
 *  console.log("Message: " + message);
 *  // INSERT Unmarshalling or Processing here
 * });
 * 
 * @example
 * <caption>An ACK is sent back to the server</caption>
 *  MSH|^~\&|SOMELAB|SOMELAB|SOMELAB|SOMELAB|20080511103530||ORU^R01|Q335939501T337311002|P|2.3|||
 *  MSA|AA|Q335939501T337311002
 * 
 */

interface MllpEvents {
    'hl7': (message: string) => void
}

export class MllpServer extends TypedEmitter<MllpEvents> {
    port: number
    host: string
    logger: any
    server: net.Server

    static readonly VT = String.fromCharCode(0x0b);
    static readonly FS = String.fromCharCode(0x1c);
    static readonly CR = String.fromCharCode(0x0d);

    constructor(host?: string, port?: number, logger?: any) {
        super();

        this.host = host || '127.0.0.1'
        this.port = port || 6969
        this.logger = logger || console.log;

        this.server = net.createServer(this.serverListener.bind(this))

    }

    private serverListener(sock: net.Socket) {
        this.logger('CONNECTED: ' + sock.remoteAddress + ':' + sock.remotePort);

        sock.on('data',  (data: any) => {
            let { msg, ack } = this.handleData(data.toString())
            
            /**
             * MLLP HL7 Event. Fired when a HL7 Message is received.
             * @event MllpServer#hl7
             * @type {string}
             * @property {string} message string containing the HL7 Message (see example below)
             * @example MSH|^~\&|XXXX|C|SOMELAB|SOMELAB|20080511103530||ORU^R01|Q335939501T337311002|P|2.3|||
             */
             this.emit('hl7', msg);

             sock.write(MllpServer.VT + this.ackn(ack, "AA") + MllpServer.FS + MllpServer.CR);
        })
        sock.on('close',  (data) => {
            this.logger('CLOSED: ' + sock.remoteAddress + ' ' + sock.remotePort);
        });

    }

    private ackn(data: any, ack_type: string): string {
        //get message ID
        var msg_id = data[0][10];

        var header = [data[0]];

        //switch around sender/receiver names
        header[0][3] = data[0][5];
        header[0][4] = data[0][6];
        header[0][5] = data[0][3];
        header[0][6] = data[0][4];

        var result = hl7.serializeJSON(header);
        result = result + "\r" + "MSA|" + ack_type + "|" + msg_id;

        return result;
    }

    private handleData(msg: string) {
        let message: string = ''
        let ackData: any

        //strip separators
        if (msg.indexOf(MllpServer.VT) > -1) {
            message = '';
        }

        message += msg.replace(MllpServer.VT, '');
        
        if (msg.indexOf(MllpServer.FS + MllpServer.CR) > -1) {
            message = message.replace(MllpServer.FS + MllpServer.CR, '');
            ackData = hl7.parseString(message);
            this.logger("Message:\r\n" + message + "\r\n\r\n");
        }

        return { msg: message, ack: ackData };
    }

    send(receivingHost: string, receivingPort: number, hl7Data: any, callback: Function) {
        const sendingClient = net.createConnection({
            host: receivingHost,
            port: receivingPort
        }, () => {
            this.logger('Sending data to ' + receivingHost + ':' + receivingPort);
            sendingClient.write(MllpServer.VT + hl7Data + MllpServer.FS + MllpServer.CR);
        });

        let _terminate = () => {
            this.logger('closing connection with ' + receivingHost + ':' + receivingPort);
            sendingClient.end();
        };
        sendingClient.on('data', (rawAckData) => {
            this.logger(receivingHost + ':' + receivingPort + ' ACKED data');

            var ackData = rawAckData
                .toString() // Buffer -> String
                .replace(MllpServer.VT, '')
                .split('\r')[1] // Ack data
                .replace(MllpServer.FS, '')
                .replace(MllpServer.CR, '');

            callback(null, ackData);
            _terminate();
        });
        sendingClient.on('error', (error) => {
            this.logger(receivingHost + ':' + receivingPort + ' couldn\'t process data');

            callback(error, null);
            _terminate();
        });
    }

    listen(callback?: Function) {
        if(callback)
            this.server.listen(this.port, this.host, () => { callback() })
        else
            this.server.listen(this.port, this.host)
    }

    close(callback: Function) {
        this.server.close(() => {
            callback()
        })
    }
}


index.js
var net = require('net');
var hl7 = require('hl7');
var util = require('util');
var EventEmitter = require('events').EventEmitter;

//The header is a vertical tab character <VT> its hex value is 0x0b.
//The trailer is a field separator character <FS> (hex 0x1c) immediately followed by a carriage return <CR> (hex 0x0d)

var VT = String.fromCharCode(0x0b);
var FS = String.fromCharCode(0x1c);
var CR = String.fromCharCode(0x0d);

/**
 * @constructor MLLPServer
 * @param {string} host a resolvable hostname or IP Address
 * @param {integer} port a valid free port for the server to listen on.
 * @param {object} logger
 * 
 * @fires MLLPServer#hl7  
 * 
 * @example
 * var server = new MLLPServer('hl7server.mydomain', 3333, console.log);
 * 
 * server.on('hl7', function(message) {
 *  console.log("Message: " + message);
 *  // INSERT Unmarshalling or Processing here
 * });
 * 
 * @example
 * <caption>An ACK is sent back to the server</caption>
 *  MSH|^~\&|SOMELAB|SOMELAB|SOMELAB|SOMELAB|20080511103530||ORU^R01|Q335939501T337311002|P|2.3|||
 *  MSA|AA|Q335939501T337311002
 * 
 */
function MLLPServer(host, port, logger) {

    var self = this;
    this.message = '';
    var HOST = host || '127.0.0.1';
    var PORT = port || 6969;
    logger = logger || console.log;

    var Server = net.createServer(function (sock) {

        logger('CONNECTED: ' + sock.remoteAddress + ':' + sock.remotePort);

        function ackn(data, ack_type) {
            //get message ID
            var msg_id = data[0][10];

            var header = [data[0]];

            //switch around sender/receiver names
            var receiverApp = data[0][5];
            var receiverInst = data[0][6];
            var senderApp = data[0][3];
            var senderInst = data[0][4];

            header[0][3] = receiverApp;
            header[0][4] = receiverInst;
            header[0][5] = senderApp;
            header[0][6] = senderInst;

            var result = hl7.serializeJSON(header);
            result = result + "\r" + "MSA|" + ack_type + "|" + msg_id;

            return result;
        }

        sock.on('data', function (data) {
            data = data.toString();
            //strip separators
            logger("DATA:\nfrom " + sock.remoteAddress + ':\n' + data.split("\r").join("\n"));

            if (data.indexOf(VT) > -1) {
                self.message = '';
            }

            self.message += data.replace(VT, '');

            if (data.indexOf(FS + CR) > -1) {
                self.message = self.message.replace(FS + CR, '');
                var data2 = hl7.parseString(self.message);
                logger("Message:\r\n" + self.message + "\r\n\r\n");
                /**
                 * MLLP HL7 Event. Fired when a HL7 Message is received.
                 * @event MLLPServer#hl7
                 * @type {string}
                 * @property {string} message string containing the HL7 Message (see example below)
                 * @example MSH|^~\&|XXXX|C|SOMELAB|SOMELAB|20080511103530||ORU^R01|Q335939501T337311002|P|2.3|||
                 */
                self.emit('hl7', self.message);
                var ack = ackn(data2, "AA");
                sock.write(VT + ack + FS + CR);
            }

        });

        sock.on('close', function (data) {
            logger('CLOSED: ' + sock.remoteAddress + ' ' + sock.remotePort);
        });

    });

    self.send = function (receivingHost, receivingPort, hl7Data, callback) {
        var sendingClient = new net.connect({
            host: receivingHost,
            port: receivingPort
        }, function () {
            logger('Sending data to ' + receivingHost + ':' + receivingPort);
            sendingClient.write(VT + hl7Data + FS + CR);
        });

        var _terminate = function () {
            logger('closing connection with ' + receivingHost + ':' + receivingPort);
            sendingClient.end();
        };

        sendingClient.on('data', function (rawAckData) {
            logger(receivingHost + ':' + receivingPort + ' ACKED data');

            var ackData = rawAckData
                .toString() // Buffer -> String
                .replace(VT, '')
                .split('\r')[1] // Ack data
                .replace(FS, '')
                .replace(CR, '');

            callback(null, ackData);
            _terminate();
        });

        sendingClient.on('error', function (error) {
            logger(receivingHost + ':' + receivingPort + ' couldn\'t process data');

            callback(error, null);
            _terminate();
        });
    };

    Server.listen(PORT, HOST);
}

util.inherits(MLLPServer, EventEmitter);

exports.MLLPServer = MLLPServer;
